demos/TicTacToe/Classes/TicTacToeScene.cpp

```diff
@@ -37,51 +282,211 @@ bool TicTacToe::init() {
+        LogMessage("Attempt to initialize Firebase Auth.");
+        void** targets = reinterpret_cast<void**>(data);
+        ::firebase::InitResult result;
+        *reinterpret_cast<::firebase::auth::Auth**>(targets[0]) =
+            ::firebase::auth::Auth::GetAuth(app, &result);
+        return result;
+      },
+      [](::firebase::App* app, void* data) {
+        LogMessage("Attempt to initialize Firebase Database.");
+        void** targets = reinterpret_cast<void**>(data);
+        ::firebase::InitResult result;
+        *reinterpret_cast<::firebase::database::Database**>(targets[1]) =
+            ::firebase::database::Database::GetInstance(app, &result);
+        return result;
+      }};
 
-  // TODO(grantpostma): This should reflect the size that is set in AppDelegate.
-  // (GetVisableSize) Should modify kTileWidth and kTileHeight based on that
-  // size. auto kScreenWidth = Director::getInstance()->getWinSize().width; auto
-  // kScreenHeight =  Director::getInstance()->getWinSize().height;
+  initializer.Initialize(app, initialize_targets, initializers,
+                         sizeof(initializers) / sizeof(initializers[0]));
+
+  WaitForCompletion(initializer.InitializeLastResult(), "Initialize");
+
+  if (initializer.InitializeLastResult().error() != 0) {
+    LogMessage("Failed to initialize Firebase libraries: %s",
+               initializer.InitializeLastResult().error_message());
+    ProcessEvents(2000);
+    return 1;
+  }
+  LogMessage("Successfully initialized Firebase Auth and Firebase Database.");
+
+  database->set_persistence_enabled(true);
+
+  // Sign in using Auth before accessing the database.
+  // The default Database permissions allow anonymous users access. This will
+  // work as long as your project's Authentication permissions allow anonymous
+  // signin.
+  {
+    firebase::Future<firebase::auth::User*> sign_in_future =
+        auth->SignInAnonymously();
+    WaitForCompletion(sign_in_future, "SignInAnonymously");
+    if (sign_in_future.error() == firebase::auth::kAuthErrorNone) {
+      LogMessage("Auth: Signed in anonymously.");
+    } else {
+      LogMessage("ERROR: Could not sign in anonymously. Error %d: %s",
+                 sign_in_future.error(), sign_in_future.error_message());
+      LogMessage(
+          "  Ensure your application has the Anonymous sign-in provider "
+          "enabled in Firebase Console.");
+      LogMessage(
+          "  Attempting to connect to the database anyway. This may fail "
+          "depending on the security settings.");
+    }
+  }
+  // Splits on the if depending on if the player created or joined the game.
+  // Additionally sets the player_index and total_players based on joining or
+  // creating a game.
+  if (join_game_uuid == NULL) {
+    join_game_uuid = random_string(4);
+    ref = database->GetReference("game_data").Child(join_game_uuid);
+    firebase::Future<void> fCreateGame = ref.Child("total_players").SetValue(1);
```

Comment:
Somewhere in the code (doesn't have to be here) can you describe the schema of the database? Just put some comments describe how you're structuring the database data, so I can more easily understand how the fields are organized and what the values represent.